File: data-structures/queue/queue.go
package queue

type Queue[T any] struct {
	data     []T
	back     uint
	front    uint
	capacity uint
}

func NewQueue[T any](capacity uint) *Queue[T] {
	return &Queue[T]{
		data:     make([]T, capacity),
		back:     0,
		front:    0,
		capacity: capacity,
	}
}

func (q Queue[T]) IsEmpty() bool {
	return q.back == q.front
}

func (q Queue[T]) IsFull() bool {
	return q.Size() == q.capacity
}

func (q Queue[T]) Size() uint {
	return q.back - q.front
}

func (q *Queue[T]) Enqueue(value T) bool {
	if q.IsFull() {
		return false
	}

	// back is at the last index and there is available space at the start of the
	// queue, shift all element towards the front.
	if q.back == (q.capacity) && q.front != 0 {
		size := q.Size()
		for i := range size {
			q.data[i] = q.data[i+q.front]
		}
		q.front = 0
		q.back = size
	}

	q.data[q.back] = value
	q.back++
	return true
}

func (q *Queue[T]) Dequeue() *T {
	if q.IsEmpty() {
		return nil
	}

	value := q.data[q.front]
	q.front++

	if q.IsEmpty() {
		q.back = 0
		q.front = 0
	}

	return &value
}

func (q Queue[T]) Peek() *T {
	if q.IsEmpty() {
		return nil
	}

	value := q.data[q.front]
	return &value
}

File: app/data_structures/queue/queue.py
from typing import Generator


class Queue[T]:
    __slots__ = ["front", "back", "capacity", "data"]

    front: int
    back: int
    capacity: int
    data: list[T | None]

    def __init__(self, capacity=5) -> None:
        self.front = self.back = 0
        self.capacity = capacity
        self.data = [None] * capacity

    def __len__(self) -> int:
        return self.back - self.front

    def is_empty(self) -> bool:
        return len(self) == 0

    def is_full(self) -> bool:
        return len(self) == self.capacity

    def __iter__(self) -> Generator[T]:
        for i in range(len(self)):
            value = self.data[i]
            if value is None:
                raise Exception(f"unexpected None value in data at index {i}")

            yield value

    def enqueue(self, value: T) -> None:
        if self.back < self.capacity:
            self.data[self.back] = value
            self.back += 1
            return

        # if there is space at the start of the queue, shift all elements forward.
        if self.front != 0:
            for i in range(len(self)):
                self.data[i] = self.data[i + self.front]

            self.back -= self.front
            self.front = 0
            self.data[self.back] = value
            self.back += 1
            return

        # resize operation.
        new_capacity = self.capacity * 2
        new_data: list[T | None] = [None] * new_capacity
        for i in range(self.front, self.back):
            new_data[i] = self.data[self.front + i]

        new_data[self.back] = value
        self.back += 1
        self.capacity = new_capacity
        self.data = new_data

    def dequeue(self) -> T | None:
        if self.is_empty():
            return None

        value = self.data[self.front]
        self.front += 1
        return value

    def peek(self) -> T | None:
        if self.is_empty():
            return None

        return self.data[self.front]
